Rust/cargo-azsphere/src/config/metadata.rs

```diff
@@ -0,0 +1,412 @@
+                    .map_err(|e| FileAnnotatedError(annot, e))?;
+                Ok(Self(table.clone(), source.clone()))
+            }
+        }
+    }
+
+    fn convert_toml_txt_to_table<'a>(
+        toml: &'a Value,
+        branch: &Option<String>,
+    ) -> Result<&'a Table, ConfigError> {
+        let root = toml
+            .as_table()
+            .ok_or(ConfigError::WrongType(".".to_string(), "table"))?;
+
+        if let Some(branch) = branch {
+            toml_dotted_bare_key_parser::parse_dotted_bare_keys(branch.as_ref())
+                .map_err(|e| ConfigError::WrongBranchPathOfToml(branch.clone(), e))?
+                .iter()
+                .fold(Some(root), |table, key| {
+                    table.and_then(|v| v.get(*key).and_then(|v| v.as_table()))
+                })
+                .ok_or(ConfigError::BranchPathNotFoundInToml(branch.to_string()))
+        } else {
+            Ok(root)
+        }
+    }
+}
+
+pub(super) struct MetadataConfig<'a> {
+    metadata: &'a Table,
+    branch_path: Option<String>,
+}
+
+impl<'a> MetadataConfig<'a> {
+    pub fn new(metadata: &'a Table, branch_path: Option<String>) -> Self {
+        Self {
+            metadata,
+            branch_path: branch_path.map(|v| v.to_string()),
+        }
+    }
+
+    pub fn new_from_extra_metadata(extra_metadata: &'a ExtraMetaData) -> Self {
+        Self::new(
+            &extra_metadata.0,
+            match &extra_metadata.1 {
+                ExtraMetadataSource::File(_, Some(branch)) => Some(branch.clone()),
+                _ => None,
+            },
+        )
+    }
+
+    pub fn new_from_manifest(manifest: &'a Manifest) -> Result<Option<Self>, Error> {
+        let pkg = manifest
+            .package
+            .as_ref()
+            .ok_or(ConfigError::Missing("package".to_string()))?;
+        let metadata = pkg.metadata.as_ref();
+        if metadata.is_none() {
+            return Ok(None);
+        }
+
+        let metadata = metadata.unwrap().as_table().ok_or(ConfigError::WrongType(
+            "package.metadata".to_string(),
+            "table",
+        ))?;
+        let metadata = metadata
+            .iter()
+            .find(|(name, _)| name.as_str() == "azsphere")
+            .ok_or(ConfigError::Missing(
+                "package.metadata.azsphere".to_string(),
+            ))?
+            .1
+            .as_table()
+            .ok_or(ConfigError::WrongType(
+                "package.metadata.azsphere".to_string(),
+                "table",
+            ))?;
+
+        Ok(Some(Self {
+            metadata: metadata,
+            branch_path: Some("package.metadata.azsphere".to_string()),
+        }))
+    }
+
+    fn create_config_error(&self, name: &str, type_name: &'static str) -> ConfigError {
+        let toml_path = self
+            .branch_path
+            .as_ref()
+            .map(|v| [v, name].join("."))
+            .unwrap_or(name.to_string());
```

Comment:
You're calling `name.to_string` every time which is an unnecessary allocation. It's preferable to call `unwrap_or_else`. Granted this isn't a very hot code path so in practice, it won't make a difference. 

I only mention this because I believe, clippy would flag this. Are we running clippy in CI?

Reply:
Bug 107